File: backend/app/analysis/confluence.py
"""Confluence scorer — the heart of the signal engine.

Implements the algorithm from SIGNAL_ENGINE.md §3 exactly.
"""

from dataclasses import dataclass
from decimal import Decimal

import pandas as pd

from app.analysis.indicators.adx import adx_factor, adx_is_strong, adx_is_weak
from app.analysis.indicators.bbands import bbands_factor
from app.analysis.indicators.ema import (
    ema_cross_factor,
    multibagger_ema_factor,
    price_vs_ema_factor,
)
from app.analysis.indicators.macd import macd_cross_factor, macd_histogram_factor
from app.analysis.indicators.rsi import rsi_divergence_factor, rsi_level_factor
from app.analysis.indicators.volume import volume_factor
from app.analysis.patterns.multi import (
    detect_engulfing,
    detect_harami,
    detect_morning_evening_star,
    detect_piercing_dark_cloud,
)
from app.analysis.patterns.single import (
    detect_hammer,
    detect_hanging_man,
    detect_marubozu,
    detect_shooting_star,
)
from app.analysis.structure.dow import dow_trend_factor
from app.analysis.structure.fibonacci import fibonacci_factor
from app.analysis.structure.institutional import fii_dii_factor
from app.analysis.structure.levels import sr_zone_factor
from app.analysis.types import FactorResult


@dataclass
class ConfluenceResult:
    direction: str          # 'BUY' | 'SELL'
    confidence_pct: int     # 0-100
    normalized_score: float # -1 to +1
    factors: list[FactorResult]
    triggering_patterns: list[str]
    triggering_indicators: list[str]
    is_multibagger: bool = False


def _best_pattern_factor(candles: pd.DataFrame, swing_n: int = 3) -> FactorResult:
    """Evaluate all patterns; return the single highest-magnitude result."""
    close = float(candles["close"].iloc[-1])
    window = candles.iloc[-20:] if len(candles) >= 20 else candles
    high_20 = float(window["high"].max())
    low_20 = float(window["low"].min())
    at_swing_low = abs(close - low_20) / low_20 <= 0.01 if low_20 > 0 else False
    at_swing_high = abs(close - high_20) / high_20 <= 0.01 if high_20 > 0 else False

    candidates = [
        detect_marubozu(candles),
        detect_hammer(candles, at_swing_low=at_swing_low),
        detect_hanging_man(candles, at_swing_high=at_swing_high),
        detect_shooting_star(candles, at_swing_high=at_swing_high),
        detect_engulfing(candles),
        detect_harami(candles),
        detect_piercing_dark_cloud(candles),
        detect_morning_evening_star(candles),
    ]

    best = max(candidates, key=lambda p: abs(p.score))
    tags = ["pattern", "bullish" if best.score > 0 else "bearish"] if best.detected else ["pattern"]
    return FactorResult(
        name=best.name,
        weight=15,
        score=best.score if best.detected else 0.0,
        explanation=best.explanation,
        tags=tags,
    )


def run_all_factors(
    candles: pd.DataFrame,
    timeframe: str = "1d",
    fii_net_5d: Decimal = Decimal("0"),
    dii_net_5d: Decimal = Decimal("0"),
    stock_block_deal_net_cr: Decimal = Decimal("0"),
) -> list[FactorResult]:
    """Run every factor and return the raw list of FactorResult objects."""
    swing_n = 3 if timeframe in ("1m", "5m", "15m") else 5

    close = float(candles["close"].iloc[-1])

    pattern_factor = _best_pattern_factor(candles, swing_n)
    bullish_pattern = pattern_factor.score > 0
    bearish_pattern = pattern_factor.score < 0

    vol_factor = volume_factor(candles)
    breakout_volume = vol_factor.score > 0

    factors: list[FactorResult] = [
        pattern_factor,
        dow_trend_factor(candles, lookback=20, swing_n=swing_n),
        ema_cross_factor(candles),
        price_vs_ema_factor(candles),
        rsi_level_factor(candles),
        rsi_divergence_factor(candles),
        macd_cross_factor(candles),
        macd_histogram_factor(candles),
        vol_factor,
        bbands_factor(candles),
        adx_factor(candles),
        sr_zone_factor(
            candles,
            current_price=close,
            bullish_pattern=bullish_pattern,
            bearish_pattern=bearish_pattern,
            breakout_volume_ok=breakout_volume,
        ),
        fibonacci_factor(candles, swing_n=swing_n),
        fii_dii_factor(fii_net_5d, dii_net_5d, stock_block_deal_net_cr),
    ]

    # Multibagger bonus (1d only)
    if timeframe == "1d":
        mb = multibagger_ema_factor(candles)
        if mb.score > 0:
            factors.append(mb)

    return factors


def score_from_factors(
    factors: list[FactorResult],
    candles: pd.DataFrame,
    min_confidence: int = 70,
) -> ConfluenceResult | None:
    """Score from a pre-built (and optionally weight-overridden) factor list.

    Separated from score_signal so BacktestEngine can inject custom weights.
    """
    total_weighted = sum(f.weight * f.score for f in factors)
    total_weight = sum(f.weight for f in factors if f.score != 0.0)

    if total_weight == 0:
        return None

    normalized = total_weighted / total_weight
    confidence_pct = int(abs(normalized) * 100)

    effective_min = min_confidence
    if adx_is_weak(candles):
        effective_min += 5
    elif adx_is_strong(candles):
        effective_min = max(65, min_confidence - 5)

    if confidence_pct < effective_min:
        return None

    direction = "BUY" if normalized > 0 else "SELL"

    triggering_patterns = [
        f.name for f in factors
        if "pattern" in f.tags and f.score != 0.0
    ]
    triggering_indicators = [
        f.name for f in factors
        if "indicator" in f.tags and f.score != 0.0
    ]

    is_multibagger = any(f.name == "MULTIBAGGER_EMA" and f.score > 0 for f in factors)

    return ConfluenceResult(
        direction=direction,
        confidence_pct=confidence_pct,
        normalized_score=normalized,
        factors=factors,
        triggering_patterns=triggering_patterns,
        triggering_indicators=triggering_indicators,
        is_multibagger=is_multibagger,
    )


def score_signal(
    candles: pd.DataFrame,
    timeframe: str = "1d",
    min_confidence: int = 70,
    fii_net_5d: Decimal = Decimal("0"),
    dii_net_5d: Decimal = Decimal("0"),
    stock_block_deal_net_cr: Decimal = Decimal("0"),
) -> ConfluenceResult | None:
    """Main entry point: run all factors and return a ConfluenceResult or None."""
    factors = run_all_factors(
        candles, timeframe, fii_net_5d, dii_net_5d, stock_block_deal_net_cr
    )
    return score_from_factors(factors, candles, min_confidence)

File: engine/crates/engine-core/src/confluence.rs
//! Confluence scorer — faithful port of backend/app/analysis/confluence.py
//! (SIGNAL_ENGINE.md §3). Replication contract (.claude/rules/rust.md):
//! zero-score exclusion from the weight denominator, `int()` TRUNCATION of
//! confidence, multibagger appended only when positive (1d), ADX regime
//! adjustment of the gate, pattern ties resolved first-max like Python.

use crate::factors::{self, Factor};
use crate::patterns as pat;
use crate::structure;
use crate::types::Bar;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Direction {
    Buy,
    Sell,
}

#[derive(Debug, Clone)]
pub struct ConfluenceOutcome {
    pub direction: Direction,
    pub confidence_pct: i32,
    pub normalized: f64,
    pub factors: Vec<Factor>,
    pub is_multibagger: bool,
}

/// Best single pattern (§2 weights table: one pattern, weight 15, not
/// summed). Context: swing proximity = close within 1% of the 20-bar
/// high/low (NOT pivot-based — Python parity).
fn best_pattern_factor(bars: &[Bar]) -> Factor {
    let Some(last) = bars.last() else {
        return Factor::grouped("PATTERN", 15.0, 0.0, "pattern");
    };
    let close = last.close;
    let start = bars.len().saturating_sub(20);
    let window = bars.get(start..).unwrap_or(bars);
    let high20 = window
        .iter()
        .map(|b| b.high)
        .fold(f64::NEG_INFINITY, f64::max);
    let low20 = window.iter().map(|b| b.low).fold(f64::INFINITY, f64::min);
    let at_swing_low = low20 > 0.0 && (close - low20).abs() / low20 <= 0.01;
    let at_swing_high = high20 > 0.0 && (close - high20).abs() / high20 <= 0.01;

    let candidates = [
        pat::detect_marubozu(bars),
        pat::detect_hammer(bars, at_swing_low),
        pat::detect_hanging_man(bars, at_swing_high),
        pat::detect_shooting_star(bars, at_swing_high),
        pat::detect_engulfing(bars),
        pat::detect_harami(bars),
        pat::detect_piercing_dark_cloud(bars),
        pat::detect_morning_evening_star(bars),
    ];
    // Python max(): first-encountered max wins ties (strict > when scanning).
    let mut best = candidates
        .first()
        .copied()
        .unwrap_or(crate::types::PatternHit::miss("PATTERN"));
    for c in candidates.iter().skip(1) {
        if c.score.abs() > best.score.abs() {
            best = *c;
        }
    }
    Factor {
        name: best.code,
        weight: 15.0,
        score: if best.detected { best.score } else { 0.0 },
        is_pattern: true,
        is_indicator: false,
        group: "pattern",
    }
}

/// Institutional inputs (crore INR) — zeros in backtests today.
#[derive(Debug, Clone, Copy, Default)]
pub struct FlowInputs {
    pub fii_net_5d: f64,
    pub dii_net_5d: f64,
    pub stock_block_deal_net_cr: f64,
}

pub fn run_all_factors(bars: &[Bar], timeframe: &str, flows: FlowInputs) -> Vec<Factor> {
    let swing_n = if matches!(timeframe, "1m" | "5m" | "15m") {
        3
    } else {
        5
    };
    let close = bars.last().map(|b| b.close).unwrap_or(f64::NAN);

    let pattern = best_pattern_factor(bars);
    let bullish_pattern = pattern.score > 0.0;
    let bearish_pattern = pattern.score < 0.0;

    let vol = factors::volume_factor(bars);
    let breakout_volume = vol.score > 0.0;

    let mut out = vec![
        pattern,
        Factor {
            name: "DOW_TREND",
            weight: 20.0,
            score: structure::dow_trend_score(bars, 20, swing_n),
            is_pattern: false,
            is_indicator: false,
            group: "structure",
        },
        factors::ema_cross_factor(bars),
        factors::price_vs_ema_factor(bars),
        factors::rsi_level_factor(bars),
        factors::rsi_divergence_factor(bars, 10),
        factors::macd_cross_factor(bars),
        factors::macd_histogram_factor(bars),
        vol,
        factors::bbands_factor(bars),
        factors::adx_factor(bars),
        Factor {
            name: "SR_ZONE",
            weight: 10.0,
            score: structure::sr_zone_score(
                bars,
                Some(close),
                bullish_pattern,
                bearish_pattern,
                breakout_volume,
            ),
            is_pattern: false,
            is_indicator: false,
            group: "structure",
        },
        Factor {
            name: "FIBONACCI",
            weight: 5.0,
            score: structure::fibonacci_score(bars),
            is_pattern: false,
            is_indicator: false,
            group: "structure",
        },
        factors::fii_dii_factor(
            flows.fii_net_5d,
            flows.dii_net_5d,
            flows.stock_block_deal_net_cr,
        ),
    ];

    if timeframe == "1d" {
        let mb = factors::multibagger_ema_factor(bars);
        if mb.score > 0.0 {
            out.push(mb);
        }
    }
    out
}

/// §3 scorer over a pre-built factor list (weights may be overridden by
/// the strategy lab before this call).
pub fn score_from_factors(
    factors: &[Factor],
    bars: &[Bar],
    min_confidence: i32,
) -> Option<ConfluenceOutcome> {
    // Adjudicated 2026-07-04 (item A): volume "only counts if direction
    // matches" — a surge confirms the OTHER factors, never opposes them,
    // never fires alone. Mirrors Python exactly.
    let rest: f64 = factors
        .iter()
        .filter(|f| f.name != "VOLUME")
        .map(|f| f.weight * f.score)
        .sum();
    let factors: Vec<Factor> = factors
        .iter()
        .map(|f| {
            if f.name == "VOLUME" && f.score > 0.0 {
                let new_score = if rest > 0.0 {
                    0.5
                } else if rest < 0.0 {
                    -0.5
                } else {
                    0.0
                };
                Factor {
                    score: new_score,
                    ..*f
                }
            } else {
                *f
            }
        })
        .collect();
    let factors = factors.as_slice();

    let total_weighted: f64 = factors.iter().map(|f| f.weight * f.score).sum();
    let total_weight: f64 = factors
        .iter()
        .filter(|f| f.score != 0.0)
        .map(|f| f.weight)
        .sum();
    if total_weight == 0.0 {
        return None;
    }
    let normalized = total_weighted / total_weight;
    #[allow(clippy::cast_possible_truncation)]
    let confidence_pct = (normalized.abs() * 100.0).trunc() as i32; // Python int()

    let mut effective_min = min_confidence;
    if factors::adx_is_weak(bars) {
        effective_min += 5;
    } else if factors::adx_is_strong(bars) {
        // Python: max(65, min_confidence - 5)
        effective_min = (min_confidence - 5).max(65);
    }
    if confidence_pct < effective_min {
        return None;
    }

    let direction = if normalized > 0.0 {
        Direction::Buy
    } else {
        Direction::Sell
    };
    let is_multibagger = factors
        .iter()
        .any(|f| f.name == "MULTIBAGGER_EMA" && f.score > 0.0);

    Some(ConfluenceOutcome {
        direction,
        confidence_pct,
        normalized,
        factors: factors.to_vec(),
        is_multibagger,
    })
}

pub fn score_signal(
    bars: &[Bar],
    timeframe: &str,
    min_confidence: i32,
    flows: FlowInputs,
) -> Option<ConfluenceOutcome> {
    let factors = run_all_factors(bars, timeframe, flows);
    score_from_factors(&factors, bars, min_confidence)
}
